interfaces/cython/cantera/examples/reactors/ic_engine.py

```diff
@@ -54,200 +56,203 @@
 injector_open = 350. / 180. * np.pi
 injector_close = 365. / 180. * np.pi
 injector_mass = 3.2e-5  # kg
-injector_t_open = (injector_close - injector_open) / 2. / np.pi / f
 
 # Simulation time and resolution
-sim_n_revolutions = 8.
-sim_n_timesteps = 100000.
+sim_n_revolutions = 8
+delta_T_max = 20.
+
+# ###################################################################
+
+# set up IC engine parameters
+V_oT = V_H / (epsilon - 1.)
+A_piston = .25 * np.pi * d_piston ** 2
+stroke = V_H / A_piston
+
+# convert time to crank angle
+
+
+def crank_angle(t):
+    return np.remainder(2 * np.pi * f * t, 4 * np.pi)
 
-###################################################################
 
 # load reaction mechanism
 gas = ct.Solution(reaction_mechanism)
 
-# define initial state
+# define initial state and set up reactor
 gas.TPX = T_inlet, p_inlet, comp_inlet
-r = ct.IdealGasReactor(gas)
+cyl = ct.IdealGasReactor(gas)
+cyl.volume = V_oT
+
 # define inlet state
 gas.TPX = T_inlet, p_inlet, comp_inlet
 inlet = ct.Reservoir(gas)
+
+# inlet valve
+inlet_valve = ct.Valve(inlet, cyl)
+inlet_delta = np.mod(inlet_close - inlet_open, 4 * np.pi)
+inlet_valve.valve_coeff = inlet_valve_coeff
+inlet_valve.set_time_function(
+    lambda t: np.mod(crank_angle(t) - inlet_open, 4 * np.pi) < inlet_delta)
+
 # define injector state (gaseous!)
 gas.TPX = T_injector, p_injector, comp_injector
 injector = ct.Reservoir(gas)
+
+# injector is modeled as a mass flow controller
+injector_mfc = ct.MassFlowController(injector, cyl)
+injector_delta = np.mod(injector_close - injector_open, 4 * np.pi)
+injector_t_open = (injector_close - injector_open) / 2. / np.pi / f
+injector_mfc.mass_flow_coeff = injector_mass / injector_t_open
+injector_mfc.set_time_function(
+    lambda t: np.mod(crank_angle(t) - injector_open, 4 * np.pi) < injector_delta)
+
 # define outlet pressure (temperature and composition don't matter)
 gas.TPX = T_ambient, p_outlet, comp_ambient
 outlet = ct.Reservoir(gas)
+
+# outlet valve
+outlet_valve = ct.Valve(cyl, outlet)
+outlet_delta = np.mod(outlet_close - outlet_open, 4 * np.pi)
+outlet_valve.valve_coeff = outlet_valve_coeff
+outlet_valve.set_time_function(
+    lambda t: np.mod(crank_angle(t) - outlet_open, 4 * np.pi) < outlet_delta)
+
 # define ambient pressure (temperature and composition don't matter)
 gas.TPX = T_ambient, p_ambient, comp_ambient
 ambient_air = ct.Reservoir(gas)
 
-# set up connecting devices
-inlet_valve = ct.Valve(inlet, r)
-injector_mfc = ct.MassFlowController(injector, r)
-outlet_valve = ct.Valve(r, outlet)
-piston = ct.Wall(ambient_air, r)
+# piston is modeled as a moving wall with sinusoidal velocity profile
+piston = ct.Wall(ambient_air, cyl)
+piston.area = A_piston
+
+
+def piston_speed(t): return - stroke / 2 * 2 * \
```

Comment:
If this has to be on two lines anyways, why not put the whole statement on one line? Also, can you add a docstring?